sqlakeyset/serial/serial.py

```diff
@@ -148,29 +167,34 @@ def unserialize_values(self, s: str) -> Optional[Tuple]:
             return None
         return tuple(self.unserialize_value(_) for _ in self.split(s))
 
-    def serialize_value(self, x) -> str:
-        try:
-            serializer = self.serializers[type(x)]
-        except KeyError:
-            pass  # fall through to builtins
-        else:
-            try:
-                c, x = serializer(x)
-            except Exception as e:
-                raise PageSerializationError(
-                    "Custom bookmark serializer " "encountered error"
-                ) from e
-            else:
-                return "{}:{}".format(c, x)
+    def get_serializer(self, x):
+        for cls in type(x).__mro__:
+            with suppress(KeyError):
+                return self.serializers[cls]
 
-        try:
+        return None
+
+    def serialize_value(self, x) -> str:
+        with suppress(KeyError):
```

Comment:
I see why you've moved the builtins to before the type serializers (to avoid the int serializer handling true/false). Unfortunately I think this slightly changes the behaviour of the public API - if anyone had defined a custom `bool` serializer it will now no longer work. Unlikely but possible... I'll have a think about whether we should care.

Reply:
I thought about it a bunch and couldn't think of a reason to care- it makes the code significantly more simple.

If you think this is over-complicating things- it might be simpler to have type registration accept multiple types to treat the same, and have a canonical version which things get serialized to.

With this, we could add the asyncpg classes to the registration (wrapped in try except in case its not installed) and we wouldn't have to add this mro complexity. Up to you, I really appreciate the review and the library!

Reply:
My first thought is that it makes more sense to handle any subclass of a registered type than to add special cases for third-party library types.

That said, after some more thought it's not immediately obvious that the substitution principle should apply at all, since we need to deserialize as well as serialize. The fact that your `asyncpg...UUID` would come back as a `uuid.UUID` after a serialize/deserialize round-trip could be very surprising! Maybe it's actually better to require the user to register the type separately than to silently convert it to the base class.

Reply:
My expectation here would be that the liskov substitution principle applies- i.e. if UUID is registered with a serializer, any subclass of UUID can be serialized with it and naturally it would get deserialized with the registered class.

Reply:
Another option is something like:

```python
custom_bookmark_type(
    AsyncPgUUID,
    "uuid",
    replace=True,
)
```

So I can not have to do `uuid2`.

Reply:
> Another option is something like:
> 
> ```python
> custom_bookmark_type(
>     AsyncPgUUID,
>     "uuid",
>     replace=True,
> )
> ```

Yeah, I definitely agree we should allow overriding the default bookmark types.

Reply:
> My expectation here would be that the Liskov Substitution Principle applies- i.e. if UUID is registered with a serializer, any subclass of UUID can be serialized with it and naturally it would get deserialized with the registered class. Since the base class is registered its sort of expected that it least has some functional value to sqlalchemy.

Mmm, you're probably right. I think I'm okay with it.